alf/algorithms/misc_algorithm.py

```diff
@@ -0,0 +1,239 @@
+                obs_achieved_goal_2 = task_specific_ob_2
+                obs_excludes_goal = agent_pose_1
+
+                return (obs_excludes_goal, obs_achieved_goal_1,
+                        obs_achieved_goal_2)
+            else:
+                print(
+                    'Please specify the state of interests and the context state in the misc_alorighm.py first.'
+                )
+                exit()
+        else:
+            print(
+                'Please specify the state of interests and the context state in the misc_alorighm.py first.'
+            )
+            exit()
+
+    def mine(self, x_in, y_in):
+
+        # Mutual Information Neural Estimation
+        y_in_tran = tf.transpose(y_in, perm=[1, 0, 2])
+        y_shuffle_tran = tf.gather(
+            y_in_tran, tf.random.shuffle(tf.range(tf.shape(y_in_tran)[0])))
+        y_shuffle = tf.transpose(y_shuffle_tran, perm=[1, 0, 2])
+        x_conc = tf.concat([x_in, x_in], axis=-2)
+        y_conc = tf.concat([y_in, y_shuffle], axis=-2)
+
+        # propagate the forward pass
+        layerx = self._misc_layerx(inputs=x_conc)
+        layery = self._misc_layery(inputs=y_conc)
+        layer2 = tf.nn.relu(layerx + layery)
+        output = self._misc_layero(inputs=layer2)
+        output = tf.nn.tanh(output)
+
+        # split in T_xy and T_x_y predictions
+        N_samples = tf.shape(x_in)[-2]
+        T_xy = output[:, :N_samples, :]
+        T_x_y = output[:, N_samples:, :]
+
+        # compute the negative loss (maximise loss == minimise -loss)
+        mean_exp_T_x_y = tf.reduce_mean(tf.math.exp(T_x_y), axis=-2)
+        loss = tf.reduce_mean(T_xy, axis=-2) - tf.math.log(mean_exp_T_x_y)
+        loss = tf.squeeze(loss)  # Mutual Information
+
+        return loss
+
+    def train_step(self, inputs, state, calc_intrinsic_reward=True):
+        """
+        Args:
+            inputs (tuple): observation
+            state (Tensor):  state for MISC (previous feature)
+            calc_intrinsic_reward (bool): if False, only return the losses
+        Returns:
+            TrainStep:
+                outputs: empty tuple ()
+                state:  empty tuple ()
+                info: (MISCInfo):
+        """
+        feature_state, prev_action = inputs
+        feature = tf.concat([feature_state, prev_action], axis=-1)
+        prev_feature = state
+
+        feature_reshaped = tf.expand_dims(feature, axis=1)
+        prev_feature_reshaped = tf.expand_dims(prev_feature, axis=1)
+        feature_pair = tf.concat([prev_feature_reshaped, feature_reshaped], -2)
+        feature_reshaped_tran = tf.transpose(feature_reshaped, perm=[1, 0, 2])
+
+        def add_batch():
+            self._buffer.add_batch(feature_reshaped_tran)
+
+        run_if(
+            tf.reduce_all(
+                tf.equal(
+                    tf.shape(feature),
+                    tf.constant(self._traj_spec.shape.as_list()))), add_batch)
+
+        if self.n_objects < 2:
```

Comment:
Generally I suggesting splitting the several use cases into different classes, probably driving from the same base class.

Reply:
Maybe, we do that later.